cip/1.accepted/CIP2021-08-10-Operator-precedence.adoc

```diff
@@ -0,0 +1,725 @@
+        5 =(5 = 5) AS c
+----
+|
+[options="header"]
+!====
+! a    ! ax   ! b     ! c
+! true ! true ! false ! false
+!====
+
+|
+[source, cypher]
+----
+RETURN  5 = 4 <> 3  AS a,
+       (5 = 4) AND (4 <> 3) AS ax,
+       (5 = 4)<> 3  AS b,
+        5 =(4 <> 3) AS c
+----
+|
+[options="header"]
+!====
+! a     ! ax    ! b    ! c
+! false ! false ! true ! false
+!====
+
+|
+[source, cypher]
+----
+RETURN  5 <> 4 > 3  AS a,
+       (5 <> 4) AND (4 > 3) AS ax,
+       (5 <> 4)> 3  AS b,
+        5 <>(4 > 3) AS c
+----
+|
+[options="header"]
+!====
+! a    ! ax   ! b    ! c
+! true ! true ! null ! true
+!====
+
+|
+[source, cypher]
+----
+RETURN  5 > 4 > 3  AS a,
+       (5 > 4) AND (4 > 3) AS ax,
+       (5 > 4)> 3  AS b,
+        5 >(4 > 3) AS c
+----
+|
+[options="header"]
+!====
+! a    ! ax   ! b    ! c
+! true ! true ! null ! null
+!====
+
+|
+[source, cypher]
+----
+RETURN  5 > 4 >= 3  AS a,
+       (5 > 4) AND (4 >= 3) AS ax,
+       (5 > 4)>= 3  AS b,
+        5 >(4 >= 3) AS c
+----
+|
+[options="header"]
+!====
+! a    ! ax   ! b    ! c
+! true ! true ! null ! null
+!====
+
+|
+[source, cypher]
+----
+RETURN  5 <= 4 < 3  AS a,
+       (5 <= 4) AND (4 < 3) AS ax,
+       (5 <= 4)< 3  AS b,
+        5 <=(4 < 3) AS c
+----
+|
+[options="header"]
+!====
+! a     ! ax    ! b    ! c
+! false ! false ! null ! null
+!====
+
+|====
+
+== What others do
+
+openCypher has grammatically all operators in one precedence level hierarchy.
+This is similar to how precedence is defined for more programming languages.
+For instance:
+
+* https://docs.oracle.com/javase/tutorial/java/nutsandbolts/operators.html[Java]
+* https://docs.python.org/3/reference/expressions.html#operator-precedence[Python]
+* https://en.cppreference.com/w/cpp/language/operator_precedence[C++]
+* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table[Javascript]
+
+The precedence is similar.
+Some minor difference:
+
+* Programming languages have extra operators openCypher does not have, e.g. increment/decrement, bitwise operators, ternary operator, and assignments
+* In Javascript exponentiation is right-to-left associative
+
+SQL encodes large parts of the type compatibility of expressions in the grammar.
+Hence, its precedence hierarchy is partition by type.
+Within one type, SQL's precedence hierarchy is similar.
+
+Chainable operators, such as addition, multiplication, etc. are left-to-right associative in SQL as well.
+However, SQL directly encode that in the grammar, with left-recursive production rules, e.g.
+
+[source, ebnf]
+----
+<numeric value expression> ::=
+    <term>
+  | <numeric value expression> <plus sign> <term>
+  | <numeric value expression> <minus sign> <term>
+
+<term> ::=
+    <factor>
+  | <term> <asterisk> <factor>
+  | <term> <solidus> <factor>
+
+<factor> ::=
+  [ <sign> ] <numeric primary>
+----
+
+Most program and query language do not give chains of comparison operators extra semantics.
```

Comment:
The most popular programming languages and query languages, do not give chains of comparison operators extra semantics.